packages/cmf/src/mergeModules.js

```diff
@@ -0,0 +1,188 @@
+
+	if (preReducer && newPreReducer) {
+		let safePreReducer = preReducer;
+		if (typeof preReducer === 'function') {
+			safePreReducer = [preReducer];
+		}
+		if (Array.isArray(newPreReducer)) {
+			return [...safePreReducer, ...newPreReducer];
+		} else if (typeof newPreReducer === 'function') {
+			return safePreReducer.concat([newPreReducer]);
+		}
+		return preReducer;
+	}
+	if (newPreReducer) {
+		return newPreReducer;
+	}
+	return preReducer;
+}
+
+function mergeEnhancer(enhancer, newEnhancer) {
+	assertTypeOf({ enhancer }, 'enhancer', 'function');
+	assertTypeOf({ enhancer: newEnhancer }, 'enhancer', 'function');
+
+	if (enhancer && newEnhancer) {
+		return mergeFns(enhancer, newEnhancer);
+	}
+	if (newEnhancer) {
+		return newEnhancer;
+	}
+	return enhancer;
+}
+
+function mergeMiddlewares(middlewares, newMiddlewares) {
+	assertTypeOf({ middlewares }, 'middlewares', 'Array');
+	assertTypeOf({ middlewares: newMiddlewares }, 'middlewares', 'Array');
+
+	if (newMiddlewares && newMiddlewares.length) {
+		return (middlewares || []).concat(newMiddlewares);
+	}
+	return middlewares;
+}
+
+function mergeStoreCallback(storeCallback, newStoreCallback) {
+	assertTypeOf({ storeCallback }, 'storeCallback', 'function');
+	assertTypeOf({ storeCallback: newStoreCallback }, 'storeCallback', 'function');
+
+	if (storeCallback && newStoreCallback) {
+		return mergeFns(storeCallback, newStoreCallback);
+	}
+	return getFirstOf(storeCallback, newStoreCallback, 'storeCallback');
+}
+
+function mergeReducer(reducer, newReducer) {
+	assertTypeOf({ reducer }, 'reducer', ['object', 'function']);
+	assertTypeOf({ reducer: newReducer }, 'reducer', ['object', 'function']);
+
+	if (reducer && newReducer) {
+		if (typeof reducer === typeof newReducer && typeof reducer === 'object') {
+			return mergeObjects(reducer, newReducer, 'reducer');
+		}
+		if (typeof reducer === 'object' && typeof newReducer === 'function') {
+			return mergeObjects(reducer, { app: newReducer }, 'reducer');
+		}
+		if (typeof newReducer === 'object' && typeof reducer === 'function') {
+			return mergeObjects({ app: reducer }, newReducer, 'reducer');
+		}
+		// both are functions
+		return mergeFns(reducer, newReducer);
+	}
+	if (newReducer) {
+		return newReducer;
+	}
+	return reducer;
+}
+
+const MERGE_FNS = {
+	appId: getFirstOf,
+	history: getFirstOf,
+	AppLoader: getFirstOf,
+	saga: mergeSaga,
+	httpMiddleware: getFirstOf,
+	preReducer: mergePreReducer,
+	enhancer: mergeEnhancer,
+	middlewares: mergeMiddlewares,
+	storeCallback: mergeStoreCallback,
+	reducer: mergeReducer,
+	preloadedState: getFirstOf,
+	settingsURL: getFirstOf,
+	registry: getReduceRegistry('registry', 'object'),
+	sagas: getReduceRegistry('sagas', 'object'),
+	components: getReduceRegistry('components', 'object'),
+	expressions: getReduceRegistry('expressions', 'object'),
+	actionCreators: getReduceRegistry('actionCreators', 'object'),
+	sagaRouterConfig: getReduceRegistry('sagaRouterConfig', 'object'),
```

Comment:
You don't need this `getReduceRegistry`. It's only objects.

In general, for each argument, you don't need all those. You have only thos cases : 
* getFirstOf --> perhaps rename to `getUnique()`
* mergeObject : if the result should be an object
* mergeArray : if the result should be an array. Use [], concat to avoid too many if to check the type (array, function, ...)
* mergeFunction : if the result should be a function that calls all the functions you pass.
* mergeSaga : special case for saga